src/MOI_wrapper.jl

```diff
@@ -0,0 +1,355 @@
+}
+
+"""
+    struct VectorizedManifold{M} <: MOI.AbstractVectorSet
+        manifold::M
+    end
+
+Representation of points of `manifold` as a vector of `R^n` where `n` is
+`MOI.dimension(VectorizedManifold(manifold))`.
+"""
+struct VectorizedManifold{M} <: MOI.AbstractVectorSet
+    manifold::M
+end
+
+function MOI.dimension(set::VectorizedManifold)
+    return prod(ManifoldsBase.representation_size(set.manifold))
+end
+
+struct _EmptyNLPEvaluator <: MOI.AbstractNLPEvaluator end
+MOI.features_available(::_EmptyNLPEvaluator) = [:Grad, :Jac, :Hess]
+MOI.initialize(::_EmptyNLPEvaluator, ::Any) = nothing
+
+"""
+    Manopt.Optimizer()
+
+Creates a new optimizer object.
+
+The minimization of a function `f(X)` of an an array `X[1:n1,1:n2,...]`
+over a manifold `M` starting at `X0`, can be modeled as follows:
+```julia
+using JuMP
+@variable(model, X[i1=1:n1,i2=1:n2,...] in M, start = X0[i1,i2,...])
+@objective(model, Min, f(X))
+```
+"""
+mutable struct Optimizer <: MOI.AbstractOptimizer
+    manifold::Union{Nothing,ManifoldsBase.AbstractManifold}
+    problem::Union{Nothing,AbstractManoptProblem}
+    state::Union{Nothing,AbstractManoptSolverState}
+    variable_primal_start::Vector{Union{Nothing,Float64}}
+    sense::MOI.OptimizationSense
+    nlp_data::MOI.NLPBlockData
+    qp_data::QPBlockData{Float64}
+    options::Dict{String,Any}
+    function Optimizer()
+        return new(
+            nothing,
+            nothing,
+            nothing,
+            Union{Nothing,Float64}[],
+            MOI.FEASIBILITY_SENSE,
+            MOI.NLPBlockData([], _EmptyNLPEvaluator(), false),
+            QPBlockData{Float64}(),
+            Dict{String,Any}(),
+        )
+    end
+end
+
+MOI.get(::Optimizer, ::MOI.SolverVersion) = "0.4.25"
+
+function MOI.is_empty(model::Optimizer)
+    return isnothing(model.manifold) &&
+           isempty(model.variable_primal_start) &&
+           model.nlp_data.evaluator isa _EmptyNLPEvaluator &&
+           model.sense == MOI.FEASIBILITY_SENSE
+end
+
+function MOI.empty!(model::Optimizer)
+    model.manifold = nothing
+    model.problem = nothing
+    model.state = nothing
+    empty!(model.variable_primal_start)
+    model.sense = MOI.FEASIBILITY_SENSE
+    model.nlp_data = MOI.NLPBlockData([], _EmptyNLPEvaluator(), false)
+    model.qp_data = QPBlockData{Float64}()
+    return nothing
+end
+
+function MOI.supports(::Optimizer, ::MOI.RawOptimizerAttribute)
+    # FIXME It should depend on `attr.name`
+    return true
+end
+
+function MOI.get(model::Optimizer, attr::MOI.RawOptimizerAttribute)
+    if !MOI.supports(model, attr)
+        throw(MOI.UnsupportedAttribute(attr))
+    end
+    return model.options[attr.name]
+end
+
+function MOI.set(model::Optimizer, attr::MOI.RawOptimizerAttribute, value)
+    if !MOI.supports(model, attr)
+        throw(MOI.UnsupportedAttribute(attr))
+    end
+    model.options[attr.name] = value
+    return nothing
+end
+
+MOI.get(::Optimizer, ::MOI.SolverName) = "Manopt"
```

Comment:
show of a state would call the summary which is maybe a bit long, but might it be reasonable to at least provide the states main type (without parameters)? That might be nice to be a bit more precise. so for example to something like `Manopt.GradientDescentState` or so – or is that misleading?
Depends of course also what this usually returns - just the package or the actual algorithm as a name as well.

Reply:
Yes, if the algorithm is specified, it's a nice addition. Let's maybe leave it as follow up PR ?